OneCommunityGlobal/HighestGoodNetworkApp
# Gary add materials page

# Description
Adds **Purchase Request: Material** page and form to the BM Dashboard (Phase 2).

## Related PRS (if any):
[Backend PR 577](https://github.com/OneCommunityGlobal/HGNRest/pull/577) must be running to test this branch.

## Main changes explained:
- adds ability to initiate a purchase request for a material to be used on a specific project. 
- request must include quantity and priority level, with an optional 'preferred brand' field.
- adds actions and reducers to store project and material types data in redux store
- adds error handling to fetch request on page load
- adds success and error toasts after submitting purchase request

## How to test:
1. follow checkout instructions for [backend PR 577](https://github.com/OneCommunityGlobal/HGNRest/pull/577)
2. `git checkout gary_add_materials_page`
3. `npm install`
4. `npm run start:local`
5. log in to HGN as a volunteer user 
6. navigate to `http://localhost:3000/bmdashboard` and enter credentials again
7. navigate to `http://localhost:3000/bmdashboard/materials/purchase`
8. Test basic form submission:
    - Project and Material options should auto-load into respective fields
    - Enter info into all required fields (Preferred Brand is optional) and click **Submit**
    - Note whether a success toast notification is received
    - Optional: check Network tab in browser tools to check for a 'materials' request with status code 201 (Created).
9. Test form validation:
    - Test that **Submit** button is disabled unless all required fields are entered
    - Test that entering a decimal or technically-allowable letter 'e' (ex: 10e10) in **Quantity** field results in a validation check/error.
    - Note: there are validation checks for all form fields, but these are harder to test without manipulating the HTML.
 10. Test that **Cancel** button clears form and returns user to previously visited page.
 11. Test error handling:
    - To test error handling on initial page load, you can deactivate the backend server and reload the page. This will load an error page with code 503. 
    - You can also change the fetch route in **src/actions/bmdashboard/projectActions.js > fetchBMProjects** from `axios.get(ENDPOINTS.BM_PROJECTS)` to `axios.get(ENDPOINTS.BM_PROJECTS + 's')`, then reload the page, to force a 404 error.
    - To test error handling if form submission fails, change the fetch route in **src/actions/bmdashboard/materialsActions > purchaseMaterial** from `ENDPOINTS.BM_MATERIALS` to `ENDPOINTS.BM_MATERIALS + "s"`. This should result in an error toast with code 404.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/98129080/6df64404-733a-422f-9e43-4cff72235fc0

## Notes:
- **Please remember to either Approve or Request Changes. Do not just leave a comment.**
- **Please leave a review for BOTH front and backend branches after you have finished testing!**
- Gary began work on this branch and has since left the project. The branch and pull request are currently being managed by Tim.
- Permissions are currently not being built into the Phase 2 codebase. All functionality should be accessible to volunteer users.
- The form will eventually only be accessible to Building Manager users. Purchase requests will be limited to the BM's own projects. For now these limitations are not in place (the code has been written but is disabled).
- Lint fixes were made to unrelated files while Gary was working on the branch: ProjectTable.jsx, reports.css, SummaryBar.jsx, UserProfileAdd.jsx, UserProfileAdd.scss.
